### ISSUE TITLE
Handle econnrefused

### ISSUE THREAD

**Issue by pankajmyt** [mysql-otp/mysql-otp#210]
Using it for few years now, recently moved to gcp. That does a bi-weekly db backup and for about a minute the db is unavailable giving econnrefused. 
This mysql+poolboy goes into a retry loop, and kills the server. What would be a good way to have some kind of reconnect in x seconds timer?

**Comment by pankajmyt** [comment]
I have not gone through internals of mysql-otp-hnc, but if connection refusal happens what's is its retry strategy?

**Comment by zuiderkwast** [comment]
It kills the server??? Or it kills the Erlang supervision tree?

Since mysql-otp itself doesn't do reconnects, this falls kind of outside the main functionality. The poolboy wrapper is just an example. I think a wrapper process that does a backoff before starting a new instance would be good. @juhlig can HNC do this already?

**Comment by pankajmyt** [comment]
Server depends on mysql_poolboy. Pool is started using mysql_poolboy:add_pool. When db restarts it does 10 failed connection attempts in 1 second, exceeding restart for worker. that takes out mysql_poolboy. the db failover takes about 1-2 minutes.

So my option seem

1. either use heart - but that has its own complications
2. put reconnection logic inside the mysql connection itself.

**Comment by zuiderkwast** [comment]
Ah, when you said it kills the server, I thought you meant it killed the MySQL server. :)

You can create a small gen server that starts the mysql process and restarts it if it dies, with some backoff time. Then you add this gen server to poolboy instead of mysql directly. I've done it in a  previous job but it was not open sourced.

Consder mysql_poolboy an example app. Write your own is better.

I would also be open to adding optionall automatic reconnect logic to mysql-otp, but the problem is that state is lost when you reconnect. If you create all the connection state (prepared statements, set variables, etc.) in the options, then it may be fine to reconnect.

**Comment by juhlig** [comment]
> @juhlig can HNC do this already?

No, but IIRC itself HNC won't crash in such a case. That said, we haven't been working on HNC for a while, because overall interest is low and my own immediate use case is gone.

> What would be a good way to have some kind of reconnect in x seconds timer?

Would using the option `connect_mode => lazy` (instead of the default `synchronous`) be an alternative? With that, `mysql` won't try to connect (and crash) immediately when started (or restarted by `poolboy`), instead it will try to connect (and crash) only when it is actually used for the first time.

**Comment by zuiderkwast** [comment]
> ... and probably very serious: open transactions; After an undetected reconnect in the middle of a transaction, all uncommitted changes will be lost, and all statements issued in the new connection will be executed outside of a transaction.

Yeah, this can be very serious. We can't let that pass undetected. We may find some a way to fail the transaction though and return 'aborted'. If we do, then we have provided a way around all the reasons for using state and then I believe we can allow reconnects. Not by default though.

**Comment by juhlig** [comment]
> We may find some a way to fail the transaction though and return 'aborted'.

For transactions executed via `mysql:transaction/2,3,4`, we probably can. For transactions started via an explicit `BEGIN` or `START TRANSACTION` statement, we probably can't :man_shrugging:

**Comment by zuiderkwast** [comment]
> For transactions executed via `mysql:transaction/2,3,4`, we probably can. For transactions started via an explicit `BEGIN` or `START TRANSACTION` statement, we probably can't 🤷‍♂️

Yes, so we can provide a way to use transactions with reconnects.

You can mess up by doing things the other way. The same is true for declaring prepared statements in the startup options vs explicit `mysql:prepare/N` or explicit `PREPARE`, or setting server-side variables in the options vs explicitly. That's why reconnect needs to be off by default, if we add it.

The documentation about the reconnect option should warn about these things you can't do if you enable it.

**Comment by dong50252409** [comment]
I am also troubled by such problems. I think a reliable solution is to add `erlang:process_flag(trap_exit, true)` in [mysql_conn:init/1](https://github.com/mysql-otp/mysql-otp/blob/1afd59677b3e2daf795f5fd77268e1dbc20f4c51/src/mysql_conn.erl#L68).
Make it truly effective [timeout](https://github.com/mysql-otp/mysql-otp/blob/1afd59677b3e2daf795f5fd77268e1dbc20f4c51/src/mysql_conn.erl#L162) during connection, so that I can achieve the desired effect by adjusting the startup policy of `mysql_poolboy_sup.erl`.

**Comment by dong50252409** [comment]
Why must spawn_link be used  [here](https://github.com/mysql-otp/mysql-otp/blob/1afd59677b3e2daf795f5fd77268e1dbc20f4c51/src/mysql_conn.erl#L145)  instead of spawn

**Comment by zuiderkwast** [comment]
I guess we could do retries, even exponential backoff, if the initial connect fails. (This is not the same as reconnect.)

@juhlig WDYT?

**Comment by juhlig** [comment]
Hm, I don't think I'm a huge fan of this... Handling retries is the responsibility of the user IMO. It is however unfortunate that the "user" in this case is poolboy, ie an intermediary library which is not in the hands of the _end_ user, so handling retries is also not in the hands of the _end_ user either. But I don't think we should have extra hoops which are somewhat un-Erlang-ish (give or take) just to please poolboy.

I guess the problem with poolboy which hits you here is that, when a worker exits (in this specific case, because the connection is lost), poolboy tries to restart it (which in this specific case fails because of the `econnrefused`, because the MySQL server is down). If the restart of a single worker fails, poolboy crashes (a known issue since 2019: devinus/poolboy#122).

Have you tried using a different `connect_mode` for the mysql-otp workers?
The default is `synchronous`, which means connecting is part of the mysql init; if it fails (with `econnrefused` for example), `start_link` will return `{error, ...}`, which makes poolboy (which invariably expects `{ok, Pid}`) crash.
If you use `asynchronous` or `lazy` as `connect_mode`, the mysql-otp worker process will start (returning `{ok, Pid}`, which is what poolboy wants) but not connect as part of init. `asynchronous` will do the connect immediately afterwards (and crash if the MySQL server is still down, but that should be fine for poolboy, it will get a worker exit and try another restart, ie in a tight loop, but not crash itself). `lazy` will connect only when the connection is to be used for the first time, with the same potential result.

**Comment by zuiderkwast** [comment]
@juhlig Yeah, whenever a mysql instance is alive, it's supposed to be connected. That's the contract so far. We don't have different states such as disconnected or connecting. If we add a state like this and reject queries when we're in this state, it might break applications, but we could make this opt-in with a config.

Most other database connectors, and also http clients like gun, actually use state machines where an instance can be connected or not. What we do, exit when we can't connect, is not erlangish because it escalates up on the supervision tree. Supervisors are for fault-tolerating internal errors, not external ones such as remote servers being down or unavailable.
